File: app/src/main/java/com/o4x/musical/ui/activities/tageditor/SongTagEditorActivity.java
package com.o4x.musical.ui.activities.tageditor;

import android.content.Intent;

import androidx.annotation.NonNull;

import com.o4x.musical.R;
import com.o4x.musical.loader.SongLoader;
import com.o4x.musical.network.Models.ITunesModel;
import com.o4x.musical.ui.activities.tageditor.onlinesearch.SongSearchActivity;

import java.util.ArrayList;
import java.util.List;

public class SongTagEditorActivity extends AbsTagEditorActivity<ITunesModel.Results> {


    @Override
    protected int getContentViewLayout() {
        return R.layout.activity_song_tag_editor;
    }

    @NonNull
    @Override
    protected List<String> getSongPaths() {
        List<String> paths = new ArrayList<>(1);
        paths.add(SongLoader.getSong(this, getId()).data);
        return paths;
    }

    @Override
    protected void fillViewsWithResult(ITunesModel.Results result) {
        loadImageFromUrl(result.getBigArtworkUrl(), result.collectionName);
        if (songName != null)
            songName.setText(result.trackName);
        if (albumName != null)
            albumName.setText(result.collectionName);
        if (artistName != null)
            artistName.setText(result.artistName);
        if (genreName != null)
            genreName.setText(result.primaryGenreName);
        if (year != null)
            year.setText(result.getYear());
        if (trackNumber != null)
            trackNumber.setText(String.valueOf(result.trackNumber));
    }

    @Override
    protected void searchImageOnWeb() {
        searchWebFor(
                songName.getText().toString(),
                albumName.getText().toString(),
                artistName.getText().toString()
        );
    }

    @Override
    protected void searchOnline() {
        Intent intent = new Intent(this, SongSearchActivity.class);
        intent.putExtra(SongSearchActivity.EXTRA_SONG_NAME, tagUtil.getSongTitle());
        this.startActivityForResult(intent, SongSearchActivity.REQUEST_CODE);
    }

    @Override
    protected void getImageFromLastFM() {

    }

}

File: app/src/main/java/com/o4x/musical/ui/activities/tageditor/SongTagEditorActivity.kt
package com.o4x.musical.ui.activities.tageditor

import android.content.Intent
import android.os.Bundle
import com.o4x.musical.extensions.show
import com.o4x.musical.model.Artist
import com.o4x.musical.network.models.ITunesModel.Results
import com.o4x.musical.ui.activities.tageditor.onlinesearch.SongSearchActivity
import java.util.*

class SongTagEditorActivity : AbsTagEditorActivity<Results>() {

    override fun albumImageView() = binding.backImage
    override fun artistImageView() = binding.frontImage

    val song by lazy { repository.songById(id) }

    override fun onCreate(savedInstanceState: Bundle?) {
        album = repository.albumById(song.albumId)
        super.onCreate(savedInstanceState)
    }

    override fun showViews() {
        binding.apply {
            song.show()
            album.show()
            artist.show()
            year.show()
            genre.show()
            track.show()
            disc.show()
            lyrics.show()
        }
    }

    override fun createArtist(): Artist =
        repository.artistById(song.artistId)

    override fun createPaths(): List<String> {
        val paths: MutableList<String> = ArrayList(1)
        paths.add(song.data)
        return paths
    }

    override fun fillViewsWithResult(result: Results) {
        loadImageFromUrl(result.bigArtworkUrl, null)
        binding.song.editText?.setText(result.trackName)
        binding.album.editText?.setText(result.collectionName)
        binding.artist.editText?.setText(result.artistName)
        binding.genre.editText?.setText(result.primaryGenreName)
        binding.year.editText?.setText(result.year)
        binding.track.editText?.setText(result.trackNumber.toString())
    }

    override fun searchImageOnWeb() {
        searchWebFor(
            binding.song.editText?.text.toString(),
            binding.album.editText?.text.toString(),
            binding.artist.editText?.text.toString()
        )
    }

    override fun searchOnline() {
        val intent = Intent(this, SongSearchActivity::class.java)
        intent.putExtra(SongSearchActivity.EXTRA_SONG_NAME, tagUtil?.songTitle)
        this.startActivityForResult(intent, SongSearchActivity.REQUEST_CODE)
    }
}